lib/fsm/follow.go

```diff
@@ -65,60 +66,50 @@ func (e *PlanRolledBackEvent) isTerminalEvent() bool { return true }
 func FollowOperationPlan(ctx context.Context, getPlan GetPlanFunc) <-chan PlanEvent {
 	ch := make(chan PlanEvent, 100)
 	// Send an initial batch of events from the initial state of the plan.
-	plan, err := getPlan()
-	if err != nil {
-		logrus.WithError(err).Error("Failed to load plan.")
-	}
-	if plan != nil {
-		events := getPlanEvents(GetPlanProgress(*plan), *plan)
-		// recreate channel with enough room for all events + some room for additional events
-		ch = make(chan PlanEvent, len(events)+10)
-
-		for _, event := range events {
-			select {
-			case ch <- event:
-			default:
-				logrus.WithField("event", event).Warn("Event channel is full.")
-			}
-			if event.isTerminalEvent() {
-				return ch
-			}
-		}
+	plan, err := sendNextPlan(ctx, nil, getPlan, ch)
+	if err == nil && plan == nil {
+		// Done
+		return ch
 	}
 	// Then launch a goroutine that will be monitoring the progress.
 	go func() {
-		tickerBackoff := getFollowBackoffPolicy()
-		ticker := backoff.NewTicker(tickerBackoff)
-		defer ticker.Stop()
+		ticker := backoff.NewTicker(getFollowStepPolicy())
+		tickerC := ticker.C
+		var errorTicker *backoff.Ticker
+		defer func() {
+			if ticker != nil {
+				ticker.Stop()
+			}
+			if errorTicker != nil {
+				errorTicker.Stop()
+			}
+		}()
 		defer logrus.Info("Operation plan watcher done.")
 		for {
 			select {
-			case <-ticker.C:
-				nextPlan, err := getPlan()
-				if err != nil {
-					logrus.WithError(err).Error("Failed to reload plan.")
-					continue
-				}
-				changes, err := DiffPlan(plan, *nextPlan)
-				if err != nil {
-					logrus.WithError(err).Error("Failed to diff plans.")
-					continue
-				}
-				for _, event := range getPlanEvents(changes, *nextPlan) {
-					select {
-					case ch <- event:
-					default:
-						logrus.WithField("event", event).Warn("Event channel is full.")
-					}
-					if event.isTerminalEvent() {
+			case <-tickerC:
+				nextPlan, err := sendNextPlan(ctx, plan, getPlan, ch)
+				if err == nil {
+					if nextPlan == nil {
+						// Done
 						return
 					}
+					// Update the current plan for comparison on the next cycle and
+					// reset the backoff so the ticker keeps ticking every second
+					// as long as there are no errors.
+					plan = nextPlan
+					if ticker == nil {
+						errorTicker.Stop()
+						errorTicker = nil
+						ticker = backoff.NewTicker(getFollowStepPolicy())
+						tickerC = ticker.C
+					}
+					continue
 				}
-				// Update the current plan for comparison on the next cycle and
-				// reset the backoff so the ticker keeps ticking every second
-				// as long as there are no errors.
-				plan = nextPlan
-				tickerBackoff.Reset()
```

Comment:
This is in violation of the package documented behavior: https://github.com/cenkalti/backoff/blob/v4/ticker.go#L26-L28 hence I added another ticker for backoffs during errors.